presto/dbapi.py

```diff
@@ -236,10 +245,111 @@ def setinputsizes(self, sizes):
+    def _execute_prepare_statement(self, operation, statement_name):
+        """
+        Prepends the given `operation` with "PREPARE <statement_name> FROM" and
+        executes as a prepare statement.
+
+        :param operation: sql to be executed.
+        :param statement_name: name that will be assigned to the prepare
+            statement.
+
+        :raises presto.exceptions.FailedToObtainAddedPrepareHeader: Error raised
+            when unable to find the 'X-Presto-Added-Prepare' for the PREPARE
+            statement request.
+
+        :return: string representing the value of the 'X-Presto-Added-Prepare'
+            header.
+        """
+        sql = 'PREPARE {statement_name} FROM {operation}'.format(
+            statement_name=statement_name,
+            operation=operation
+        )
+
+        # Send prepare statement. Copy the _request object to avoid poluting the
+        # one that is going to be used to execute the actual operation.
+        query = presto.client.PrestoQuery(copy.deepcopy(self._request), sql=sql)
+        result = query.execute()
+
+        # Iterate until the 'X-Presto-Added-Prepare' header is found or
+        # until there are no more results
+        for _ in result:
+            response_headers = result.response_headers
+
+            if constants.HEADER_ADDED_PREPARE in response_headers:
+                return response_headers[constants.HEADER_ADDED_PREPARE]
+
+        raise presto.exceptions.FailedToObtainAddedPrepareHeader
+
+    def _get_added_prepare_statement_presto_query(
+        self,
+        added_prepare_header,
+        statement_name,
+        params
+    ):
+        sql = 'EXECUTE ' + statement_name + ' USING ' + ','.join(map(str, params))
+
+        # No need to deepcopy _request here because this is the actual request
+        # operation
+        return presto.client.PrestoQuery(self._request, sql=sql)
+
+    def _execute_deallocate_prepare(self, added_prepare_header, statement_name):
+        sql = 'DEALLOCATE PREPARE ' + statement_name
+
+        # Send deallocate statement. Copy the _request object to avoid poluting the
+        # one that is going to be used to execute the actual operation.
+        query = presto.client.PrestoQuery(copy.deepcopy(self._request), sql=sql)
+        result = query.execute(
+            additional_http_headers={
+                constants.HEADER_PREPARED_STATEMENT: added_prepare_header
+            }
+        )
+
+        # Iterate until the 'X-Presto-Deallocated-Prepare' header is found or
+        # until there are no more results
+        for _ in result:
+            response_headers = result.response_headers
+
+            if constants.HEADER_DEALLOCATED_PREPARE in response_headers:
+                return response_headers[constants.HEADER_DEALLOCATED_PREPARE]
+
+        raise presto.exceptions.FailedToObtainDeallocatedPrepareHeader
+
+    def execute(self, operation, params=None):
+        if params:
```

Comment:
@findepi @electrum Should we start looking at how other DB-API drivers in Python solve this by providing multiple Cursor classes and over-ride the `execute()` and related functions accordingly in the sub-classes?

Will keep things cleaner and allow adding new Cursors that can maybe return the data as list of dicts etc.